Migrate Package managers

keshav-space opened the pull request:
This PR migrates existing [package managers code in VulnerableCode ](https://github.com/nexB/vulnerablecode/blob/d3f314dce082e1b70350019137687a1d214b19f2/vulnerabilities/package_managers.py)to FetchCode, also refactor and streamline the consumption using purl router.

JonoYang reviewed (changes requested):
@keshav-space I've left some suggestions to avoid accessing a dictionary multiple times for the same key-value pair.

keshav-space commented:
Thanks @JonoYang, made changes as per your suggestions.

pombredanne commented:
LGTM... we need to find a better name for this module, may be "package_versions.py" for now?

Beyond this we need to have a better design.
https://github.com/package-url/packageurl-python/blob/main/src/packageurl/contrib/purl2url.py#L70 design is not great but has its use.

Here are some thoughts (to track in new issue(s)):
1. We are going to have discrete functions each with a router that return a very specific URL or small piece of data like today. But here the functions that create URLs and fetch should be split in two so that we can have URL-only functions as explained in 2. that do not fetch anything.

2. Functions that only transform a PURL in a URL should be in one place (likely packageurl-python)

3. Then anything that fetches remote data should be of two kinds
 - One may return raw data from a JSON or XML API
 - One may return a ScanCode Package object converted from this raw data 
 
4. Some basic function that only return versions may just return lists of PURLs alright

5. We need to account to for repository_url and download_url qualifiers


We also need to make the migration for VulnerableCode with this new code. Can you start a PR in parallel so we avoid duplicating code. In VCIO you could use a temp requirements in setup.cfg such as `git+https://github.com/nexB/fetchcode@6faf26353b8cbc65b07ba6c3285a0e8c8a1c9f1b` to collect a specific commit

pombredanne reviewed (changes requested):
Thanks. See fedback

keshav-space commented on src/fetchcode/package_managers.py:
```diff
@@ -0,0 +1,523 @@
+    """Fetch versions of ``github`` packages using GitHub REST API."""
+    purl = PackageURL.from_string(purl)
+    response = get_response(
+        url=(f"https://api.github.com/repos/{purl.namespace}/{purl.name}/releases"),
```
Use GitHub tags when release is not available. 
For example, https://github.com/nginx/nginx/releases doesn't have any release, but it does have tags https://github.com/nginx/nginx/tags.

keshav-space commented:
> LGTM... we need to find a better name for this module, may be "package_versions.py" for now?

package_managers.py is now renamed to `package_versions.py`.

> Beyond this we need to have a better design. https://github.com/package-url/packageurl-python/blob/main/src/packageurl/contrib/purl2url.py#L70 design is not great but has its use.
> 
> Here are some thoughts (to track in new issue(s)):
> 
> 1. We are going to have discrete functions each with a router that return a very specific URL or small piece of data like today. But here the functions that create URLs and fetch should be split in two so that we can have URL-only functions as explained in 2. that do not fetch anything.
> 2. Functions that only transform a PURL in a URL should be in one place (likely packageurl-python)
> 3. Then anything that fetches remote data should be of two kinds
> 
> * One may return raw data from a JSON or XML API
> * One may return a ScanCode Package object converted from this raw data
> 
> 4. Some basic function that only return versions may just return lists of PURLs alright
> 5. We need to account to for repository_url and download_url qualifiers
>

We're tracking this here: https://github.com/nexB/purldb/issues/233  


> We also need to make the migration for VulnerableCode with this new code. Can you start a PR in parallel so we avoid duplicating code. In VCIO you could use a temp requirements in setup.cfg such as `git+https://github.com/nexB/fetchcode@6faf26353b8cbc65b07ba6c3285a0e8c8a1c9f1b` to collect a specific commit

Added PR for this here: https://github.com/nexB/vulnerablecode/pull/1354

JonoYang reviewed (changes requested):
@keshav-space I left some comments regarding adding comments or docstring tests.

keshav-space commented on src/fetchcode/package_versions.py:
```diff
@@ -0,0 +1,638 @@
+
+def get_response(url, content_type="json", headers=None):
+    """Fetch ``url`` and return its content as ``content_type`` which is one of binary, text or json."""
+    assert content_type in ("binary", "text", "json", "yaml")
```
@JonoYang we can remove this assert altogether since this method is not meant to be directly accessed from outside and is just a utility function for different ecosystems to fetch the response in the appropriate format. The only way for this to go wrong is when the ecosystem function is incorrect in the first place.

JonoYang reviewed (approved):
@keshav-space I think this looks good!
